### PR TITLE
0.2.0

### PR DESCRIPTION
**Pull request by sugarshin**
@ngs お手すきでレビューお願いします 🙇

### CODE REVIEW

File: src/Toolbar.js [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+          className={classnames('rte-toolbar-button-heading', {
+            active: HEADER_BLOCK_TYPES.some(t => t === getCurrentBlockType(editorState))
+          })}
+          key={ITEM_NAMES.HEADING}
+          name={itemOptions[ITEM_NAMES.HEADING].name}
+          closeButtonNode={itemOptions[ITEM_NAMES.HEADING].closeButtonNode}
+          editorState={editorState}
+          onToggle={editorState => {
+            onChangeEditorState(editorState);
+            this.props.onToggleHeadingAfter();
```
Comment (ngs) [inline src/Toolbar.js]: We should null check for non-required function properties.

Wrapping with instance method makes easy to maintain.

File: src/Body.js [hunk 2]
```diff
@@ -45,102 +46,124 @@ export default class RichTextEditor extends Component {
...
-  _blockRendererFn(block) { // eslint-disable-line complexity
+  _blockStyleFn(block) {
+    if (typeof this.props.blockStyleFn === 'function') {
+      return this.props.blockStyleFn(block);
+    }
+    switch (block.getType()) {
+    case BLOCK_TYPES.CHECKABLE_LIST_ITEM: return BLOCK_TYPES.CHECKABLE_LIST_ITEM;
+    case OLD_BLOCK_TYPES.ALIGN_CENTER: return OLD_BLOCK_TYPES.ALIGN_CENTER;
+    case OLD_BLOCK_TYPES.ALIGN_RIGHT: return OLD_BLOCK_TYPES.ALIGN_RIGHT;
+    case OLD_BLOCK_TYPES.ALIGN_JUSTIFY: return OLD_BLOCK_TYPES.ALIGN_JUSTIFY;
```
Comment (ngs) [inline src/Body.js]: Returning same value??

Comment (ngs) [inline src/Body.js]: If you want to restrict values I suggest using more simple style:

```js
const type = block.getType()
switch (type) {
case BLOCK_TYPES.CHECKABLE_LIST_ITEM:
case OLD_BLOCK_TYPES.ALIGN_CENTER:
case OLD_BLOCK_TYPES.ALIGN_RIGHT:
case OLD_BLOCK_TYPES.ALIGN_JUSTIFY:
  return type;
default:
  return '';
}
```

Comment (sugarshin) [inline src/Body.js]: Oh I see!

**Comment by sugarshin** [comment]
@ngs pls re-review 🙇

File: src/Toolbar.js [hunk 3]
```diff
@@ -43,6 +43,12 @@ export default class Toolbar extends Component {
   }
   constructor(props) {
     super(props);
+
+    this.toggleHeadingAfter = () => this._toggleHeadingAfter();
+    this.handleClickInsertImage = ev => this._handleClickInsertImage(ev);
+    this.handleClickUploadFile = ev => this._handleClickUploadFile(ev);
+    this.handleMouseDownEmbedIFrame = ev => this._handleMouseDownEmbedIFrame(ev);
+    this.handleChangeEditorState = editorState => this.props.onChangeEditorState(editorState);
```
Comment (ngs) [inline src/Toolbar.js]: Null check needed

File: src/Toolbar.js [hunk 4]
```diff
@@ -250,4 +256,24 @@ export default class Toolbar extends Component {
       ]}</div>
     );
   }
+  _handleClickInsertImage(ev) {
+    if (typeof this.props.onClickInsertImage === 'function') {
```
Comment (ngs) [inline src/Toolbar.js]: How about using [`_.isFunction`](https://lodash.com/docs#isFunction) if we [depend on lodash](https://github.com/oneteam-dev/oneteam-rte/blob/master/package.json#L29)?

File: src/Toolbar.js [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+            <span>{(checkCurrentBlockType(editorState, BLOCK_TYPES.ORDERED_LIST_ITEM) &&
+              itemOptions[ITEM_NAMES.OERDERD_LIST].activeIconNode) ||
+              itemOptions[ITEM_NAMES.OERDERD_LIST].iconNode}</span>
+          </OverlayTrigger>
+        </OrderedList>
+      ]}</div>
+    );
+  }
+  _handleClickInsertImage(ev) {
+    if (typeof this.props.onClickInsertImage === 'function') {
```
Comment (ngs) [inline src/Toolbar.js]: How about using [`_.isFunction`](https://lodash.com/docs#isFunction) if we [depend on lodash](https://github.com/oneteam-dev/oneteam-rte/blob/master/package.json#L29)?

**Comment by ngs** [comment]
@sugarshin Can you squash `WIP` commits and change commit message easy to recognize the changes?

**Comment by sugarshin** [comment]
@ngs It was rebase commit. pls checkitout.

File: src/Body.js [hunk 6]
```diff
@@ -45,102 +46,127 @@ export default class RichTextEditor extends Component {
...
     if (!contentState.hasText()) {
       if (contentState.getBlockMap().first().getType() !== BLOCK_TYPES.UNSTYLED) {
         return true;
       }
     }
     return false;
   }
-  _blockRendererFn(block) { // eslint-disable-line complexity
+  _blockStyleFn(block) {
+    if (typeof this.props.blockStyleFn === 'function') {
```
Comment (ngs) [inline src/Body.js]: `isFunction`

File: src/Body.js [hunk 7]
```diff
@@ -45,102 +46,127 @@ export default class RichTextEditor extends Component {
...
+        checked: !!checkedState[blockKey],
+        onChangeChecked: checked => {
+          const newCheckedState = assign({}, checkedState, { [blockKey]: checked });
+          this.props.onChangeCheckedState(newCheckedState);
+        }
+      }
+    };
+  }
+  _blockRendererFn(block) {
     if (typeof this.props.blockRendererFn === 'function') {
```
Comment (ngs) [inline src/Body.js]: `isFunction`

File: src/Body.js [hunk 8]
```diff
@@ -214,18 +240,20 @@ export default class RichTextEditor extends Component {
...
       if (newEditorState !== editorState) {
-        this.onChange(newEditorState);
+        this.props.onChangeEditorState(newEditorState);
         return true;
       }
     }
+    return false;
   }
   _handlePastedFiles(files) {
     if (typeof this.props.onPastedFiles === 'function') {
```
Comment (ngs) [inline src/Body.js]: `isFunction`

File: src/Body.js [hunk 9]
```diff
@@ -45,102 +46,127 @@ export default class RichTextEditor extends Component {
 
     this.focus = () => this.refs.editor.focus();
     this.blur = () => this.refs.editor.blur();
-    this.onChange = editorState => this.props.onChange(editorState);
-    this.onChangeCheckedState = checkedState => this.props.onChangeCheckedState(checkedState);
+    this.handleClickWrapper = ev => this._handleClickWrapper(ev);
+    this.handleChangeEditor = editorState => this.props.onChangeEditorState(editorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 10]
```diff
@@ -45,102 +46,127 @@ export default class RichTextEditor extends Component {
...
+  _checkableListItemRenderer(block) {
+    const blockKey = block.getKey();
+    const { checkedState } = this.props;
+    return {
+      component: CheckableListItem,
+      props: {
+        checked: !!checkedState[blockKey],
+        onChangeChecked: checked => {
+          const newCheckedState = assign({}, checkedState, { [blockKey]: checked });
+          this.props.onChangeCheckedState(newCheckedState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 11]
```diff
@@ -157,7 +183,7 @@ export default class RichTextEditor extends Component {
         if(currentBlock.getDepth() === 0) {
           const newEditorState = removeBlockStyle(editorState);
           if(newEditorState) {
-            this.onChange(newEditorState);
+            this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 12]
```diff
@@ -169,7 +195,7 @@ export default class RichTextEditor extends Component {
             MAX_LIST_DEPTH
           );
           if(newEditorState) {
-            this.onChange(newEditorState);
+            this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 13]
```diff
@@ -179,15 +205,15 @@ export default class RichTextEditor extends Component {
       if(currentBlock.getLength() === 0 && currentBlock.getKey() === firstBlockKey) {
         const newEditorState = removeBlockStyle(editorState);
         if(newEditorState) {
-          this.onChange(newEditorState);
+          this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 14]
```diff
@@ -179,15 +205,15 @@ export default class RichTextEditor extends Component {
...
           return true;
         }
       }
     }
 
     const newEditorState = RichUtils.handleKeyCommand(editorState, command);
     if (newEditorState) {
-      this.onChange(newEditorState);
+      this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 15]
```diff
@@ -200,7 +226,7 @@ export default class RichTextEditor extends Component {
     const { editorState } = this.props;
     const newEditorState = RichUtils.onTab(ev, editorState, MAX_LIST_DEPTH);
     if (newEditorState !== editorState) {
-      this.onChange(newEditorState);
+      this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 16]
```diff
@@ -214,18 +240,20 @@ export default class RichTextEditor extends Component {
       ev.preventDefault();
       const newEditorState = insertText(editorState, '\t');
       if (newEditorState !== editorState) {
-        this.onChange(newEditorState);
+        this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 17]
```diff
@@ -258,7 +286,7 @@ export default class RichTextEditor extends Component {
         if (block.getDepth() === 0) {
           const newEditorState = removeBlockStyle(editorState);
           if (newEditorState) {
-            this.onChange(newEditorState);
+            this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 18]
```diff
@@ -270,7 +298,7 @@ export default class RichTextEditor extends Component {
             MAX_LIST_DEPTH
           );
           if (newEditorState) {
-            this.onChange(newEditorState);
+            this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/Body.js [hunk 19]
```diff
@@ -299,7 +327,7 @@ export default class RichTextEditor extends Component {
             blockKey,
             BLOCK_TYPES.UNSTYLED
           );
-          this.onChange(newEditorState);
+          this.props.onChangeEditorState(newEditorState);
```
Comment (ngs) [inline src/Body.js]: null check

File: src/RTE.js [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+import Body from './Body';
+import Toolbar from './Toolbar';
+import stateToHTML from './stateToHTML';
+import { createEditorState, createCheckedState } from './utils';
+import { getIFrameAttrs } from './helpers'
+import LinkDecorator from './decorators/LinkDecorator';
+import DownloadLinkDecorator from './decorators/DownloadLinkDecorator';
+import { ENTITY_TYPES } from './constants';
+
+export default class RTE extends Component {
```
Comment (ngs) [inline src/RTE.js]: I prefer not using acronym for public class. `RichTextEditor` would be better.

File: src/RTE.js [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+    this.onChangeEditorState = editorState => this.setState({ editorState });
+    this.onChangeCheckedState = checkedState => this.setState({ checkedState });
+    this.insertImage = imageFile => this._insertImage(imageFile);
+    this.insertDownloadLink = file => this._insertDownloadLink(file);
+    this.insertIFrame = iframeTagString => this._insertIFrame(iframeTagString);
+  }
+  render() {
+    const { editorState, checkedState } = this.state;
+    const content = Children.map(this.props.children, child => {
```
Comment (ngs) [inline src/RTE.js]: don't we need null checking here?

**Comment by sugarshin** [comment]
@ngs How does that look ?

**Comment by ngs** [comment]
Looks great to me

[LGTM](http://lgtm.in/i/Q2spT3cZp)

**Comment by sugarshin** [comment]
Thank you review 🙇

**Pull request merged**